lib/netext/httpext/request.go

```diff
@@ -66,6 +67,20 @@ func (u URL) GetURL() *url.URL {
 	return u.u
 }
 
+// Return a sanitized URL, clean of e.g. user credentials
+func cleanURL(urlString string) string {
+	idxEnd := strings.Index(urlString, "@")
+	if idxEnd == -1 {
+		return urlString
+	}
+	idxStart := strings.Index(urlString[:idxEnd], "/")
+	if idxStart != -1 {
+		return fmt.Sprintf("%s%s", urlString[0:idxStart+2], urlString[idxEnd+1:])
+	} else {
+		return urlString
+	}
+}
```

Comment:
I don't like this implementation, but it was the easiest way to clean both the `url` and `name` tags.

Initially I tried `url.Parse()`-ing, setting `User = nil` and returning `String()`, which is much cleaner, but this has the risk of mangling non-URL `name` tags, which would need escaping, etc. and it gets messy quickly as well. I also didn't make it a method on `httpext.URL`, since we use `url.URL` directly in `transport.go:114`, and that needs to be cleaned also.

Let me know if you have a better idea of doing this.

Also, note that this removes the credentials entirely, but we might want to consider masking them instead, e.g. "https://******:******@example.com/", in case that is more useful to see.

Reply:
I mean you can do:

```
if {
    return x
}

return y
```
instead of:

```
if {
    return x
else {
    return y
}
```

Reply:
Yeah, @golangcibot pointed that out as well, d'oh!

But I mean more from a design perspective, if it's fine that were parsing strings like this. It makes me feel uneasy that we might miss an edge case :)

Reply:
Yeah, I don't like this stringy approach as well, but I'm not sure how exactly you should implement something better... Though I'm also not sure that the current way (`strings.Index(urlString, "@")`) won't result in any false-positives - can't you have `@` in the query parameters for example?

I'd suggest investigating how our custom `httpext.URL` type works and is how and where it's used, because it may actually give us an opportunity to sanitize the credentials in its constructor. That way we can then use their sanitized versions when we get to constructing the metric tags. In my head, we have the following requirements for this task:
- any specified credentials shoudn't be visible in the system metric tags `url` and `name`
- however, if the user manually specifies the `name` tag, then that shouldn't ever be modified, even if it contains credentials
- any specified credentials should still be present in the `url` property of the JS `Response` object: https://docs.k6.io/docs/response-k6http
- ideally (see below), it would be good to still distinguish between metrics from authenticated and non-authenticated URLs

> Also, note that this removes the credentials entirely, but we might want to consider masking them instead, e.g. `https://******:******@example.com/`, in case that is more useful to see.

I personally like that idea very much - it will help users distinguish which metrics come from an authenticated URL without exposing the credentials.

Reply:
> can't you have @ in the query parameters for example?

I was relying on "@" being a reserved character, and that users would escape it as `%40` in the path or query string. :) But yeah, far from ideal.

I like your suggestions, thanks. I'll go through the flow of how `httpext.URL` is used again and think about a nicer solution, while incorporating the masking feature.

Reply:
@na-- take a look at the latest commit 599112f (I'll squash it before merging).

I think the implementation is more robust now, I took inspiration from `http.stripPassword()`. :)

I preferred to leave it as a standalone function instead of coupling it with `httpext.URL`, since that struct is not used everywhere, and the cleaning is only done where needed, so it doesn't change how URLs are handled internally, so less chance for bugs.

I also considered cleaning the WebSocket URL tag, and even though [the spec mentions HTTP auth as possible](https://tools.ietf.org/html/rfc6455#section-10.5), `gorilla/websocket` [doesn't support passing credentials in the URI](https://github.com/gorilla/websocket/blob/483fb8d7c32fcb4b5636cd293a92e3935932e2f4/client.go#L173-L176) (though it does support manually setting the `Authorization` header), so that wouldn't work for this use case anyway.

One small note, though: this is currently deviating from:

> however, if the user manually specifies the `name` tag, then that shouldn't ever be modified, even if it contains credentials

That is, if the name tag looks like a URL and it has credentials, it will be cleaned regardless if the user set it or not. In general, I think this is a good idea in order to protect users against inadvertently leaking credentials, but it also keeps this implementation simpler, so win=win? :)

Let me know what you think.

Reply:
> One small note, though: this is currently deviating from:
>  >  however, if the user manually specifies the name tag, then that shouldn't ever be modified, even if it contains credentials
>
> That is, if the name tag looks like a URL and it has credentials, it will be cleaned regardless if the user set it or not. In general, I think this is a good idea in order to protect users against inadvertently leaking credentials, but it also keeps this implementation simpler, so win=win? :)

Actually, I'm not sure this is currently deviating from it. Looking at the `if _, ok := tags["name"]; !ok && state.Options.SystemTags["name"]` check below, any manually specified `name` tag won't be modified. And I strongly feel that users should be able to manually specify whatever `name` tag they desire, even if it contains credentials. I feel like automagically sanitizing user-specified values would just annoy users, or at least it will annoy me as a user, given that I can just choose to set another tag k6 wouldn't care about...

Reply:
Ah, you're right, I misread and didn't test that. A manually set `name` tag doesn't get cleaned.

I see your point about it possibly being annoying, and stand corrected. :)

Let me know if there's anything pending here.